pkg/destroy/aws/shared.go

```diff
@@ -0,0 +1,233 @@
+								}
+								continue
+							}
+							arns = append(arns, arnString)
+						}
+					}
+
+					return !lastPage
+				},
+			)
+			if err != nil {
+				err = errors.Wrap(err, "get tagged resources")
+				o.Logger.Info(err)
+				nextTagClients = append(nextTagClients, tagClient)
+				continue
+			}
+			if len(arns) == 0 {
+				o.Logger.Debugf("No matches in %s for %s: shared, removing client", *tagClient.Config.Region, key)
+				continue
+			}
+			nextTagClients = append(nextTagClients, tagClient)
+
+			for i := 0; i < len(arns); i += 20 {
+				request.ResourceARNList = make([]*string, 0, 20)
+				for j := 0; i+j < len(arns) && j < 20; j++ {
+					request.ResourceARNList = append(request.ResourceARNList, aws.String(arns[i+j]))
+				}
+				_, err = tagClient.UntagResourcesWithContext(ctx, request)
+				if err != nil {
+					err = errors.Wrap(err, "untag shared resources")
+					o.Logger.Info(err)
+					continue
+				}
+				for j := 0; i+j < len(arns) && j < 20; j++ {
+					arn := arns[i+j]
+					o.Logger.WithField("arn", arn).Infof("Removed tag %s: shared", key)
+					removed[arn] = exists
+				}
+			}
+		}
+		tagClients = nextTagClients
+	}
+
+	return nil
+}
+
+func (o *ClusterUninstaller) cleanSharedARN(ctx context.Context, session *session.Session, arn arn.ARN, logger logrus.FieldLogger) error {
+	switch service := arn.Service; service {
+	case "route53":
+		return o.cleanSharedRoute53(ctx, session, arn, logger)
+	default:
+		logger.Debugf("nothing to clean for shared %s resource", service)
+		return nil
+	}
+}
+
+func (o *ClusterUninstaller) cleanSharedRoute53(ctx context.Context, session *session.Session, arn arn.ARN, logger logrus.FieldLogger) error {
+	client := route53.New(session)
+
+	resourceType, id, err := splitSlash("resource", arn.Resource)
+	if err != nil {
+		return err
+	}
+	logger = logger.WithField("id", id)
+
+	switch resourceType {
+	case "hostedzone":
+		return o.cleanSharedHostedZone(ctx, client, id, logger)
+	default:
+		logger.Debugf("nothing to clean for shared %s resource", resourceType)
+		return nil
+	}
+}
+
+func (o *ClusterUninstaller) cleanSharedHostedZone(ctx context.Context, client *route53.Route53, id string, logger logrus.FieldLogger) error {
+	if o.ClusterDomain == "" {
+		logger.Debug("no cluster domain specified in metadata; cannot clean the shared hosted zone")
+		return nil
+	}
+	dottedClusterDomain := o.ClusterDomain + "."
+
+	publicZoneID, err := findAncestorPublicRoute53(ctx, client, o.ClusterDomain, logger)
+	if err != nil {
+		return err
+	}
+
+	var lastError error
+	err = client.ListResourceRecordSetsPagesWithContext(
+		ctx,
+		&route53.ListResourceRecordSetsInput{HostedZoneId: aws.String(id)},
+		func(results *route53.ListResourceRecordSetsOutput, lastPage bool) bool {
+			for _, recordSet := range results.ResourceRecordSets {
+				// skip record sets that are not part of the cluster
+				name := aws.StringValue(recordSet.Name)
+				if !strings.HasSuffix(name, dottedClusterDomain) {
```

Comment:
Is the name of the record set dotted or is it the the value of the record that is dotted? When I look in the GUI it looks like the name isn't dotted but the value is.

Reply:
The name of the record set is dotted.
For example,
```
        {
            "Name": "api-int.ewolinetz3.devcluster.openshift.com.",
            "Type": "A",
            "AliasTarget": {
                "HostedZoneId": "ZLMOA37VPKANP",
                "DNSName": "ewolinetz3-rpcfk-int-42decb5de56b9b0b.elb.us-east-2.amazonaws.com.",
                "EvaluateTargetHealth": false
            }
        },
```